File: python-way/com/crack/snap/make/easy/array/sum_of_two.py
# @author Mohan Sharma

def two_sum_n2(nums, target) -> list | None:
	"""
	Finds two indices in the list `nums` such that their values add up to `target` using O(n^2) time complexity.

	Args:
		nums (list): List of integers to search through.
		target (int): The target sum to find.

	Returns:
		list | None: A list containing the two indices if a pair is found, otherwise None.

	Example:
		>>> two_sum_n2([2, 7, 11, 15], 9)
		[0, 1]

	Explanation:
		1. Iterate through the list `nums` using `enumerate` to get both the index `i` and the value `first_number`.
		2. For each `first_number`, iterate through the list again to get the index `j` and the value `second_number`.
		3. Check if the sum of `first_number` and `second_number` equals `target`:
			- If it does, return a list containing the indices `[i, j]`.
		4. If no pair is found after the loop, return None.
	"""
	for i, first_number in enumerate(nums):
		for j, second_number in enumerate(nums):
			if first_number + second_number == target:
				return [i, j]
	return None

def two_sum_nlogn(nums, target) -> list | None:
	"""
	Finds two indices in the list `nums` such that their values add up to `target` using O(n log n) time complexity.

	Args:
		nums (list): List of integers to search through.
		target (int): The target sum to find.

	Returns:
		list | None: A list containing the two indices if a pair is found, otherwise None.

	Example:
		>>> two_sum_nlogn([2, 7, 11, 15], 9)
		[0, 1]

	Explanation:
		1. Sort the list `nums` along with their original indices.
		2. Initialize two pointers `left` and `right`.
		3. Iterate through the list while `left` is less than `right`:
			- If the sum of the values at `left` and `right` is greater than `target`, decrement `right`.
			- If the sum is less than `target`, increment `left`.
			- If the sum equals `target`, return a list containing the original indices `[left, right]`.
		4. If no pair is found after the loop, return None.
	"""
	nums_with_indices = sorted((num, i) for i, num in enumerate(nums))
	left, right = 0, len(nums) - 1
	while left < right:
		current_sum = nums_with_indices[left][0] + nums_with_indices[right][0]
		if current_sum > target:
			right -= 1
		elif current_sum < target:
			left += 1
		else:
			return [nums_with_indices[left][1], nums_with_indices[right][1]]
	return None

def two_sum_hashing(nums, target) -> list | None:
	"""
	Finds two indices in the list `nums` such that their values add up to `target` using a hash map for O(n) time complexity.

	Args:
		nums (list): List of integers to search through.
		target (int): The target sum to find.

	Returns:
		list | None: A list containing the two indices if a pair is found, otherwise None.

	Example:
		>>> two_sum_hashing([2, 7, 11, 15], 9)
		[0, 1]

	Explanation:
		1. Initialize an empty dictionary `hash_map` to store the indices of the numbers.
		2. Iterate through the list `nums` using `enumerate` to get both the index `i` and the value `first_number`.
		3. Calculate `second_number` as the difference between `target` and `first_number`.
		4. Check if `second_number` is already in `hash_map`:
			- If it is, return a list containing the index of `second_number` from `hash_map` and the current index `i`.
		5. If `second_number` is not in `hash_map`, add `first_number` and its index `i` to `hash_map`.
		6. If no pair is found after the loop, return None.
	"""
	hash_map = {}
	for i, first_number in enumerate(nums):
		second_number = target - first_number
		if second_number in hash_map:
			return [hash_map[second_number], i]
		hash_map[first_number] = i
	return None
	
if __name__ == '__main__':
	arr = [2, 7, 11, 15]
	target = 13
	print(two_sum_n2(arr, target))
	print(two_sum_nlogn(arr, target))
	print(two_sum_hashing(arr, target))

File: java-way/src/main/java/com/crack/snap/make/easy/array/SumOfTwo.java
package com.crack.snap.make.easy.array;

import java.util.Arrays;
import java.util.HashMap;

/**
 * @author Mohan Sharma
 */
public class SumOfTwo {

    /**
     * This method finds two indices in the array such that their values add up to the target.
     * It uses a brute force approach with a time complexity of O(n^2).
     *
     * @param nums the array of integers
     * @param target the target sum
     * @return an array of two indices whose values add up to the target, or null if no such indices exist
     */
    public int[] twoSumOn2(int[] nums, int target) {
        // Iterate over each element in the array
        for (int i = 0; i < nums.length; i++) {
            // For each element, iterate over the subsequent elements
            for (int j = i + 1; j < nums.length; j++) {
                // Check if the sum of the two elements equals the target
                if (nums[i] + nums[j] == target) {
                    // If so, return the indices of these two elements
                    return new int[]{i, j};
                }
            }
        }
        // If no such elements are found, return null
        return null;
    }

    /**
     * This method finds two indices in the array such that their values add up to the target.
     * It first sorts the array and then uses a two-pointer approach with a time complexity of O(n log n).
     *
     * @param nums the array of integers
     * @param target the target sum
     * @return an array of two indices whose values add up to the target, or null if no such indices exist
     */
    public int[] twoSumNlogN(int[] nums, int target) {
        // Sort the array
        Arrays.sort(nums);
        // Initialize two pointers, one at the beginning and one at the end of the array
        int left = 0, right = nums.length - 1;
        // Iterate while the two pointers do not cross each other
        while (left < right) {
            // If the sum of the elements at the two pointers is greater than the target, move the end pointer left
            if (nums[left] + nums[right] > target) right--;
                // If the sum is less than the target, move the start pointer right
            else if (nums[left] + nums[right] < target) left++;
                // If the sum equals the target, return the indices of the two elements
            else if (nums[left] + nums[right] == target) {
                return new int[]{left, right};
            }
        }
        // If no such elements are found, return null
        return null;
    }

    /**
     * This method finds two indices in the array such that their values add up to the target.
     * It uses a HashMap to store the indices of the elements with a time complexity of O(n).
     *
     * @param nums the array of integers
     * @param target the target sum
     * @return an array of two indices whose values add up to the target, or null if no such indices exist
     */
    public int[] twoSumHashing(int[] nums, int target) {
        // Create a HashMap to store the indices of the elements
        var hash_map = new HashMap<Integer, Integer>();

        // Iterate over each element in the array
        for (int i = 0;  i < nums.length; i++) {
            // Calculate the second number needed to reach the target
            int firstNumber = nums[i];
            int secondNumber = target - firstNumber;
            // Check if the second number is already in the hash_map
            if (hash_map.containsKey(secondNumber)) {
                // If so, return the indices of the two elements
                return new int[]{hash_map.get(secondNumber), i};
            }
            // Otherwise, add the first number and its index to the hash_map
            hash_map.put(firstNumber, i);
        }
        // If no such elements are found, return null
        return null;
    }

    public static void main(String[] args) {
        int[] nums = {2, 7, 15, 11};
        int target = 13;
        var sumOfTwo = new SumOfTwo();
        System.out.println(Arrays.toString(sumOfTwo.twoSumOn2(nums, target)));
        System.out.println(Arrays.toString(sumOfTwo.twoSumNlogN(new int[]{2, 7, 15, 11}, target)));
        System.out.println(Arrays.toString(sumOfTwo.twoSumHashing(nums, target)));
    }
}
